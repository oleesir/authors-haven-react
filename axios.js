import axios from 'axios';

const instance = axios.create({
  baseURL: `${process.env.PUBLIC_API_URL}`,
});

instance.defaults.headers.post['Content-Type'] = 'application/json';

export default instance;
