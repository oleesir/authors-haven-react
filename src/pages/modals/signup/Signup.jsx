import React, { useContext, useState } from 'react';
import { connect } from 'react-redux';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import Aux from '../../../../hoc/Aux';
import Button from '../../../components/UI/Button/Button';
import Backdrop from '../../../components/UI/Backdrop/Backdrop';
import ModalContext from '../../../context/Modal/ModalContext';
import {
	signup as signupAction,
	clearErrors as clearErrorsAction,
} from '../../../redux/actions/authAction';
import classes from './Signup.module.scss';

export const Signup = ({ signup, error, clearErrors }) => {
	const modalContext = useContext(ModalContext);
	const { showModal, removeModal } = modalContext;
	const [message, setMessage] = useState('');

	const clearFormError = () => {
		removeModal();
		clearErrors();
	};

	const formik = useFormik({
		initialValues: {
			firstName: '',
			lastName: '',
			email: '',
			password: '',
		},
		validationSchema: Yup.object({
			firstName: Yup.string()
				.min(2, 'First name should be at least 2 characters')
				.matches(/^[A-Z\s]*$/i, 'First name must contain only letters')
				.required('First name is required'),
			lastName: Yup.string()
				.min(2, 'Last name should be at least 2 characters')
				.matches(/^[A-Z\s]*$/i, 'Last name must contain only letters')
				.required('Last name is required'),
			email: Yup.string()
				.email('Invalid email address')
				.required('Email is required'),
			password: Yup.string()
				.min(8, 'Password should be be at least 8 characters')
				.required('Password is required'),
		}),
		onSubmit: async (values, { setSubmitting, resetForm }) => {
			setSubmitting(true);

			await signup(
				values,
				() => {
					setMessage('Please check your email to confirm');
					setSubmitting(false);
					resetForm();
				},
				() => {
					setSubmitting(false);
				},
			);
		},
	});

	const handleChange = (event) => {
		clearErrors();
		setMessage('');
		return formik.handleChange(event);
	};

	return (
		<Aux>
			<Backdrop showModal={showModal} removeModal={clearFormError} />

			<div className={classes.signup}>
				<img
					className={classes.close}
					src='/src/images/closed.svg'
					onClick={clearFormError}
				/>
				<div className={classes.signupContent}>
					<div className={classes.Messages}>
						<div className={classes.leftMessage}>
							<p className={classes.leftMessageOne}>Share your stories</p>
							<p className={classes.leftMessageTwo}>Create your free Account</p>
							<p className={classes.leftMessageThree}>
								Sign into your account to serve you more personalized stories
								from creatives.
							</p>
						</div>
					</div>

					<div className={classes.signupForm}>
						<p className={classes.select}>Select one to get started</p>
						<div className={classes.social}>
							<a>
								<img
									src='/src/images/google.svg'
									className={classes.socialIcon}
								/>
							</a>
							<a>
								<img
									src='/src/images/twitter.svg'
									className={classes.socialIcon}
								/>
							</a>
							<a>
								<img
									src='/src/images/facebook.svg'
									className={classes.socialIcon}
								/>
							</a>
						</div>

						{message && (
							<div className={classes.success} data-testid='signupFormSuccess'>
								{message}
							</div>
						)}

						{error?.form && (
							<div data-testid='signupFormError' className={classes.errors}>
								{error?.form}
							</div>
						)}

						<form
							onSubmit={formik.handleSubmit}
							className={classes.theForm}
							data-testid='form'
						>
							<div className={classes.formContent}>
								<input
									id='firstName'
									name='firstName'
									type='text'
									placeholder='Firstname'
									onChange={handleChange}
									onBlur={formik.handleBlur}
									value={formik.values.firstName}
									data-testid='signupFieldFN'
								/>
								{formik.touched.firstName &&
									(formik.errors.firstName || error?.firstName) && (
										<span
											data-testid='signupFieldFNError'
											className={classes.errors}
										>
											{formik.errors.firstName || error?.firstName}
										</span>
									)}

								<input
									id='lastName'
									name='lastName'
									type='text'
									placeholder='Lastname'
									onChange={handleChange}
									onBlur={formik.handleBlur}
									value={formik.values.lastName}
									data-testid='signupFieldLN'
								/>
								{formik.touched.lastName &&
									(formik.errors.lastName || error?.lastName) && (
										<span
											data-testid='signupFieldLNError'
											className={classes.errors}
										>
											{formik.errors.lastName || error?.lastName}
										</span>
									)}

								<input
									id='email'
									name='email'
									type='email'
									placeholder='Email'
									onChange={handleChange}
									onBlur={formik.handleBlur}
									value={formik.values.email}
									data-testid='signupFieldEmail'
								/>
								{formik.touched.email && (formik.errors.email || error?.email) && (
									<span
										data-testid='signupFieldEmailError'
										className={classes.errors}
									>
										{formik.errors.email || error?.email}
									</span>
								)}

								<input
									id='password'
									name='password'
									type='password'
									placeholder='Password'
									onChange={handleChange}
									onBlur={formik.handleBlur}
									value={formik.values.password}
									data-testid='signupFieldPassword'
								/>
								{formik.touched.password &&
									(formik.errors.password || error?.password) && (
										<span
											data-testid='signupFieldPasswordError'
											className={classes.errors}
										>
											{formik.errors.password || error?.password}
										</span>
									)}
							</div>

							<div className={classes.signupButton}>
								<Button
									btnType='authBtn'
									types='submit'
									disabled={formik.isSubmitting}
								>
									Signup
								</Button>
							</div>
						</form>
						<p className={classes.already}>Already have an account? Login.</p>
					</div>
				</div>
			</div>
		</Aux>
	);
};

const mapStateToProps = (state) => ({
	error: state?.auth?.error,
});

const mapDispatchToProps = {
	signup: signupAction,
	clearErrors: clearErrorsAction,
};

export default connect(mapStateToProps, mapDispatchToProps)(Signup);
