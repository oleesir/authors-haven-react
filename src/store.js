import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import rootReducer from './redux/reducers';

const initialState = {};

const middleware = [thunk];

const store = createStore(
  rootReducer,
  initialState,
  composeWithDevTools(applyMiddleware(...middleware)),
);

// for tests
export const getStore = (state = initialState) => createStore(rootReducer, state, applyMiddleware(...middleware));

export default store;
