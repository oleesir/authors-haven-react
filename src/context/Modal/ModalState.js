import React, { useReducer } from 'react';
import ModalContext from './ModalContext';
import modalReducer from './ModalReducer';
import Signup from '../../pages/modals/signup/Signup.jsx';
import Login from '../../pages/modals/login/Login.jsx';
import {
	SIGNUP_MODAL,
	LOGIN_MODAL,
	SET_MODAL,
	REMOVE_MODAL,
} from './ModalType';

const ModalState = (props) => {
	const initialState = {
		showModal: false,
		modal: null,
	};

	const [state, dispatch] = useReducer(modalReducer, initialState);

	const setModal = (modal) => {
		dispatch({
			type: SET_MODAL,
			payload: modal,
		});
	};

	const removeModal = () => {
		dispatch({ type: REMOVE_MODAL });
	};

	const renderModal = (modal) => {
		switch (modal) {
			case SIGNUP_MODAL:
				return <Signup />;
			case LOGIN_MODAL:
				return <Login />;
			default:
				return null;
		}
	};

	return (
		<ModalContext.Provider
			value={{
				showModal: state.showModal,
				modal: state.modal,
				setModal,
				removeModal,
			}}
		>
			{props.children}
			{renderModal(state.modal)}
		</ModalContext.Provider>
	);
};

export default ModalState;
