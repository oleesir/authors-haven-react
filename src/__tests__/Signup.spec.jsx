import React from 'react';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom/extend-expect';
import { render, waitFor, act } from '@testing-library/react';
import { fireEvent, getByTestId } from '@testing-library/dom';
import axios from '../../axios';
import Signup from '../pages/modals/signup/Signup';
import ModalState from '../context/Modal/ModalState';
import { getStore } from '../store';

let mockPost;

jest.mock('axios', () => {
	mockPost = jest.fn();
	const realAxios = jest.requireActual('axios');

	return {
		...realAxios,
		create: jest.fn().mockImplementation((config) => ({
			...realAxios.create(config),
			post: mockPost,
		})),
		post: mockPost,
	};
});

const startingState = {
	auth: {
		user: null,
		isLoading: false,
		isAuthenticated: false,
		error: null,
	},
};

const renderWithRedux = (component, initialState = startingState) => {
	const store = getStore(initialState);

	return render(<Provider store={store}>{component}</Provider>);
};

it('renders signup form correctly', () => {
	const { container } = renderWithRedux(
		<ModalState>
			<Signup />
		</ModalState>,
	);
	expect(getByTestId(container, 'form')).toBeVisible();
	expect(getByTestId(container, 'signupFieldFN')).toBeInTheDocument();
	expect(getByTestId(container, 'signupFieldLN')).toBeInTheDocument();
	expect(getByTestId(container, 'signupFieldEmail')).toBeInTheDocument();
	expect(getByTestId(container, 'signupFieldPassword')).toBeInTheDocument();
	expect(container.firstChild).toMatchSnapshot();
});

it('should validate firstname input field', async () => {
	const { getByTestId, queryByTestId } = renderWithRedux(
		<ModalState>
			<Signup />
		</ModalState>,
	);

	const firstNameInput = getByTestId('signupFieldFN');
	expect(queryByTestId('signupFieldFNError')).toBeNull();

	await act(async () => {
		await fireEvent.change(firstNameInput, { target: { value: '' } });
		await fireEvent.blur(firstNameInput);
	});

	await waitFor(async () => {
		expect(getByTestId('signupFieldFNError')).toBeInTheDocument();
		expect(getByTestId('signupFieldFNError').innerHTML).toEqual(
			'First name is required',
		);
	});

	await act(async () => {
		await fireEvent.change(firstNameInput, { target: { value: 22 } });
		await fireEvent.blur(firstNameInput);
	});

	await waitFor(async () => {
		expect(getByTestId('signupFieldFNError')).toBeInTheDocument();
		expect(getByTestId('signupFieldFNError').innerHTML).toEqual(
			'First name must contain only letters',
		);
	});

	await act(async () => {
		await fireEvent.change(firstNameInput, { target: { value: 'N' } });
		await fireEvent.blur(firstNameInput);
	});

	await waitFor(async () => {
		expect(getByTestId('signupFieldFNError')).toBeInTheDocument();
		expect(getByTestId('signupFieldFNError').innerHTML).toEqual(
			'First name should be at least 2 characters',
		);
	});

	await act(async () => {
		await fireEvent.change(firstNameInput, { target: { value: 'Olisa' } });
		await fireEvent.blur(firstNameInput);
	});

	await waitFor(async () => {
		expect(queryByTestId('signupFieldFNError')).toBeNull();
	});
});

it('should validate lastname input field', async () => {
	const { getByTestId, queryByTestId } = renderWithRedux(
		<ModalState>
			<Signup />
		</ModalState>,
	);
	const lastNameInput = getByTestId('signupFieldLN');

	expect(queryByTestId('signupFieldLNError')).toBeNull();

	await act(async () => {
		await fireEvent.change(lastNameInput, { target: { value: '' } });
		await fireEvent.blur(lastNameInput);
	});

	await waitFor(async () => {
		expect(getByTestId('signupFieldLNError')).toBeInTheDocument();
		expect(getByTestId('signupFieldLNError').innerHTML).toEqual(
			'Last name is required',
		);
	});

	await act(async () => {
		await fireEvent.change(lastNameInput, { target: { value: 22 } });
		await fireEvent.blur(lastNameInput);
	});

	await waitFor(async () => {
		expect(getByTestId('signupFieldLNError')).toBeInTheDocument();
		expect(getByTestId('signupFieldLNError').innerHTML).toEqual(
			'Last name must contain only letters',
		);
	});

	await act(async () => {
		await fireEvent.change(lastNameInput, { target: { value: 'N' } });
		await fireEvent.blur(lastNameInput);
	});

	await waitFor(async () => {
		expect(getByTestId('signupFieldLNError')).toBeInTheDocument();
		expect(getByTestId('signupFieldLNError').innerHTML).toEqual(
			'Last name should be at least 2 characters',
		);
	});

	await act(async () => {
		await fireEvent.change(lastNameInput, { target: { value: 'Name' } });
		await fireEvent.blur(lastNameInput);
	});

	await waitFor(async () => {
		expect(queryByTestId('signupFieldLNError')).toBeNull();
	});
});

it('should validate email input field', async () => {
	const { getByTestId, queryByTestId } = renderWithRedux(
		<ModalState>
			<Signup />
		</ModalState>,
	);
	const emailInput = getByTestId('signupFieldEmail');
	expect(queryByTestId('signupFieldEmailError')).toBeNull();

	await act(async () => {
		await fireEvent.change(emailInput, { target: { value: '' } });
		await fireEvent.blur(emailInput);
	});

	await waitFor(() => {
		expect(getByTestId('signupFieldEmailError')).toBeInTheDocument();
		expect(getByTestId('signupFieldEmailError').innerHTML).toEqual(
			'Email is required',
		);
	});

	await act(async () => {
		await fireEvent.change(emailInput, { target: { value: 'help.com' } });
		await fireEvent.blur(emailInput);
	});

	await waitFor(() => {
		expect(getByTestId('signupFieldEmailError')).toBeInTheDocument();
		expect(getByTestId('signupFieldEmailError').innerHTML).toEqual(
			'Invalid email address',
		);
	});

	await act(async () => {
		await fireEvent.change(emailInput, { target: { value: 'help@gil.com' } });
		await fireEvent.blur(emailInput);
	});

	await waitFor(async () => {
		expect(queryByTestId('signupFieldEmailError')).toBeNull();
	});
});

it('should validate password input field', async () => {
	const { getByTestId, queryByTestId } = renderWithRedux(
		<ModalState>
			<Signup />
		</ModalState>,
	);
	const passwordInput = getByTestId('signupFieldPassword');
	expect(queryByTestId('signupFieldPasswordError')).toBeNull();

	await act(async () => {
		await fireEvent.change(passwordInput, { target: { value: '' } });
		await fireEvent.blur(passwordInput);
	});

	await waitFor(() => {
		expect(getByTestId('signupFieldPasswordError')).toBeInTheDocument();
		expect(getByTestId('signupFieldPasswordError').innerHTML).toEqual(
			'Password is required',
		);
	});

	await act(async () => {
		await fireEvent.change(passwordInput, { target: { value: 'qwerty' } });
		await fireEvent.blur(passwordInput);
	});

	await waitFor(() => {
		expect(getByTestId('signupFieldPasswordError')).toBeInTheDocument();
		expect(getByTestId('signupFieldPasswordError').innerHTML).toEqual(
			'Password should be be at least 8 characters',
		);
	});

	await act(async () => {
		await fireEvent.change(passwordInput, {
			target: { value: 'qwertyuiopas' },
		});
		await fireEvent.blur(passwordInput);
	});

	await waitFor(async () => {
		expect(queryByTestId('signupFieldPasswordError')).toBeNull();
	});
});

it('should submit a validated form', async () => {
	const { getByTestId, getByText, queryByTestId } = renderWithRedux(
		<ModalState>
			<Signup />
		</ModalState>,
	);

	mockPost.mockResolvedValueOnce({
		data: {
			data: {
				firstName: 'test',
				lastName: 'test',
				email: 'test@test.com',
			},
		},
	});

	const firstNameInput = getByTestId('signupFieldFN');
	const lastNameInput = getByTestId('signupFieldLN');
	const emailInput = getByTestId('signupFieldEmail');
	const passwordInput = getByTestId('signupFieldPassword');
	const submitBtn = getByText(/signup/i);

	expect(queryByTestId('signupFormSuccess')).toBeNull();

	await act(async () => {
		await fireEvent.change(firstNameInput, {
			target: { value: 'Hello' },
		});
		await fireEvent.blur(firstNameInput);

		await fireEvent.change(lastNameInput, {
			target: { value: 'World' },
		});
		await fireEvent.blur(lastNameInput);

		await fireEvent.change(emailInput, {
			target: { value: 'hello@gnail.com' },
		});
		await fireEvent.blur(emailInput);

		await fireEvent.change(passwordInput, {
			target: { value: 'Helloqwewrtty' },
		});
		await fireEvent.blur(passwordInput);

		await fireEvent.click(submitBtn);
	});

	await waitFor(() => {
		expect(queryByTestId('signupFormError')).toBeNull();
		expect(getByTestId('signupFormSuccess')).toBeInTheDocument();
		expect(getByTestId('signupFormSuccess').innerHTML).toEqual(
			'Please check your email to confirm',
		);
		expect(axios.post).toHaveBeenCalledTimes(1);

		axios.post.mockReset();
	});
});

it('should show a form error when submit fails', async () => {
	const { getByTestId, getByText, queryByTestId, container } = renderWithRedux(
		<ModalState>
			<Signup />
		</ModalState>,
	);

	mockPost.mockRejectedValueOnce({
		response: { data: { error: 'User with that email already exists' } },
	});

	const firstNameInput = getByTestId('signupFieldFN');
	const lastNameInput = getByTestId('signupFieldLN');
	const emailInput = getByTestId('signupFieldEmail');
	const passwordInput = getByTestId('signupFieldPassword');
	const submitBtn = getByText(/signup/i);

	expect(queryByTestId('signupFormError')).toBeNull();

	await act(async () => {
		await fireEvent.change(firstNameInput, {
			target: { value: 'Hello' },
		});
		await fireEvent.blur(firstNameInput);

		await fireEvent.change(lastNameInput, {
			target: { value: 'World' },
		});
		await fireEvent.blur(lastNameInput);

		await fireEvent.change(emailInput, {
			target: { value: 'hello@gnail.com' },
		});
		await fireEvent.blur(emailInput);

		await fireEvent.change(passwordInput, {
			target: { value: 'Helloqwewrtty' },
		});
		await fireEvent.blur(passwordInput);

		await fireEvent.click(submitBtn);
	});

	await waitFor(() => {
		expect(queryByTestId('signupFormSuccess')).toBeNull();
		expect(getByTestId('signupFormError')).toBeInTheDocument();
		expect(getByTestId('signupFormError').innerHTML).toEqual(
			'User with that email already exists',
		);
		expect(axios.post).toHaveBeenCalledTimes(1);
	});
});
