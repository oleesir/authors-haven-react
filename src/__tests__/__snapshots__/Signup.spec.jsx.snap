// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders signup form correctly 1`] = `
<div
  class="signup"
>
  <img
    class="close"
    src="/src/images/closed.svg"
  />
  <div
    class="signupContent"
  >
    <div
      class="Messages"
    >
      <div
        class="leftMessage"
      >
        <p
          class="leftMessageOne"
        >
          Share your stories
        </p>
        <p
          class="leftMessageTwo"
        >
          Create your free Account
        </p>
        <p
          class="leftMessageThree"
        >
          Sign into your account to serve you more personalized stories from creatives.
        </p>
      </div>
    </div>
    <div
      class="signupForm"
    >
      <p
        class="select"
      >
        Select one to get started
      </p>
      <div
        class="social"
      >
        <a>
          <img
            class="socialIcon"
            src="/src/images/google.svg"
          />
        </a>
        <a>
          <img
            class="socialIcon"
            src="/src/images/twitter.svg"
          />
        </a>
        <a>
          <img
            class="socialIcon"
            src="/src/images/facebook.svg"
          />
        </a>
      </div>
      
      <form
        class="theForm"
        data-testid="form"
      >
        <div
          class="formContent"
        >
          <input
            data-testid="signupFieldFN"
            id="firstName"
            name="firstName"
            placeholder="Firstname"
            type="text"
            value=""
          />
          <input
            data-testid="signupFieldLN"
            id="lastName"
            name="lastName"
            placeholder="Lastname"
            type="text"
            value=""
          />
          <input
            data-testid="signupFieldEmail"
            id="email"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <input
            data-testid="signupFieldPassword"
            id="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="signupButton"
        >
          <button
            class="Button authBtn"
            type="submit"
          >
            Signup
          </button>
        </div>
      </form>
      <p
        class="already"
      >
        Already have an account? Login.
      </p>
    </div>
  </div>
</div>
`;
